redis/src/cluster_async/mod.rs

```diff
@@ -1688,6 +1687,192 @@ impl Connect for MultiplexedConnection {
+/// This function takes a node's address, examines if its host has encountered a DNS change, where the node's endpoint now leads to a different IP address.
+/// If no socket addresses are discovered for the node's host address, or if it's a non-DNS address, it returns false.
+/// In case the node's host address resolves to socket addresses and none of them match the current connection's IP,
+/// a DNS change is detected, so the current connection isn't valid anymore and a new connection should be made.
+async fn is_dns_changed(addr: &str, curr_ip: &IpAddr) -> bool {
+    let (host, port) = match get_host_and_port_from_addr(addr) {
+        Some((host, port)) => (host, port),
+        None => return false,
+    };
+    let mut updated_addresses = match get_socket_addrs(host, port).await {
+        Ok(socket_addrs) => socket_addrs,
+        Err(_) => return false,
+    };
+
+    !updated_addresses.any(|socket_addr| socket_addr.ip() == *curr_ip)
+}
+
+fn warn_mismatch_ip(addr: &str, new_ip: Option<IpAddr>, prev_ip: Option<IpAddr>) {
+    warn!(
+        "New IP was found for node {:?}: 
+                new connection IP = {:?}, previous connection IP = {:?}",
+        addr, new_ip, prev_ip
+    );
+}
+
+fn warn_management_conn_faild(addr: &str, err: RedisError) {
+    warn!(
+        "Failed to create management connection for node {:?}. Error: {:?}",
+        addr, err
+    );
+}
+
+fn create_or_change_async_node<C>(
+    user_conn: C,
+    management_conn: Option<C>,
+    ip: Option<IpAddr>,
+    node: Option<AsyncClusterNode<C>>,
+) -> AsyncClusterNode<C>
+where
+    C: ConnectionLike + Connect + Send + Sync + 'static + Clone,
+{
+    let user_conn = async { user_conn }.boxed().shared();
+    let management_conn =
+        management_conn.map(|management_conn| async { management_conn }.boxed().shared());
+    let mut new_node;
+    if let Some(node) = node {
+        new_node = node;
+        new_node.user_connection = user_conn;
+        new_node.management_connection = management_conn;
+        new_node.ip = ip;
+    } else {
+        new_node = AsyncClusterNode::new(user_conn, management_conn, ip);
+    }
+    new_node
+}
+
+async fn connect_and_check_all_connections<C>(
+    addr: &str,
+    params: ClusterParams,
+    socket_addr: Option<SocketAddr>,
+    node: Option<AsyncClusterNode<C>>,
+) -> RedisResult<AsyncClusterNode<C>>
+where
+    C: ConnectionLike + Connect + Send + Sync + 'static + Clone,
+{
+    match future::join(
+        create_connection(addr, params.clone(), socket_addr),
+        create_connection(addr, params.clone(), socket_addr),
+    )
+    .await
+    {
+        (Ok(conn_1), Ok(conn_2)) => {
+            // Both connections were successfully established
+            let (mut user_conn, user_ip): (C, Option<IpAddr>) = conn_1;
+            let (mut management_conn, management_ip): (C, Option<IpAddr>) = conn_2;
+            if user_ip != management_ip && user_ip.is_some() {
```

Comment:
IMO the happy path should appear first - `if user_ip == management_ip {`

Reply:
changed